### PR TITLE
use global_run_context.params if run.assignments is empty

### PR DESCRIPTION
**Pull request by yuhong-sigopt**
#SN-257

### CODE REVIEW

**Review by taylorjacklespriggs — CHANGES_REQUESTED** [review]
I don't think this is the behavior we want. In a notebook, if I do
```python
sigopt.params.setdefaults(learning_rate=0.001)
```
then
```python
%%run
sigopt.params.log_learning_rate = 0.001.  # will raise KeyError
```

`fixed_values` are values that are immutable. I think that values set in the global scope should not be immutable after the run starts, so that experimentation can happen. There is a complicated set of circumstances that could happen here, happy to go over them with you offline.

**Comment by yuhong-sigopt** [comment]
```
class GlobalRunParameters(MutableMapping):
  def __init__(self, global_run_context):
    _set(self, "__global_run_context", global_run_context)
    _set(self, "__global_params", RunParameters(global_run_context, dict()))

  @property
  def __params(self):
    run_context = _get(self, "__global_run_context").run_context
    if run_context is None:
      return _get(self, "__global_params")
    return run_context.params
```

From above code, I think in `run`, `__global_params` is never used. 

In my change, if `run` can not get assignment, it uses `__global_params`. The code blow works,

```
import sigopt
import os
# os.environ["SIGOPT_API_TOKEN"] = #SIGOPT API TOKEN
os.environ['SIGOPT_PROJECT'] = 'run-examples'

%reload_ext sigopt

def func(n):
    return n * n

def train():
    v = func(sigopt.params.n)
    sigopt.log_metric("accuracy", v)

def a():
    sigopt.params.setdefaults(n=5)
    train()

def b():
    sigopt.params.n = 100
    train()
    
%%run
a()

%%run
b()
```

**Comment by taylorjacklespriggs** [comment]
```python
import sigopt
%load_ext sigopt
```
```python
sigopt.params.setdefaults(learning_rate=0.01)
```
```python
%%run
sigopt.params.learning_rate = 1
```
```
ValueError: value of 'learning_rate' cannot be changed
```

I don't think this is the desired behavior

**Comment by taylorjacklespriggs** [comment]
If you implement this behavior, the following should be supported:
```python
# cell 1
import sigopt
%load_ext sigopt

# cell 2
sigopt.params.setdefaults(tuned=0, not_tuned=0)

# cell 3
%%run
sigopt.params.tuned = 1
sigopt.params.not_tuned = 1
assert sigopt.params.tuned == 1
assert sigopt.params.not_tuned == 1

# cell 4
assert sigopt.params.tuned == 0
assert sigopt.params.not_tuned == 0

# cell 5
%%experiment
name: test experiment
parameters:
  - name: tuned
    type: int
    bounds:
      min: 1
      max: 2
metrics:
  - name: m1
budget: 1

# cell 6
%%optimize
assert 1 <= sigopt.params.tuned <= 2
assert sigopt.params.not_tuned == 0
try:
  sigopt.params.tuned = 0
  raise Exception
except ValueError:
  pass
sigopt.params.not_tuned = 1
assert sigopt.params.not_tuned == 1

# cell 7
assert sigopt.params.tuned == 0
assert sigopt.params.not_tuned == 0

**Comment by taylorjacklespriggs** [comment]
bonus: after a run is created, it could be modified so that it includes all of the untuned parameters in the global scope

File: sigopt/run_params.py [hunk 1]
```diff
@@ -30,7 +35,7 @@ def setdefaults(self, *args, **kwds):
     unset_keys = set(tmp) - set(items)
     update = {key: tmp[key] for key in unset_keys}
     items.update(update)
-    _get(self, "__run_context").set_parameters(update)
+    _get(self, "__run_context").set_parameters(items)
```
Comment (taylorjacklespriggs) [inline sigopt/run_params.py:38]: why this change? this uploads all of the values every time `setdefaults` is used. The existing behavior only modifies the values that are changed

Comment (yuhong-sigopt) [reply sigopt/run_params.py:38]: ok, I will change it back.

File: sigopt/run_params.py [hunk 2]
```diff
@@ -7,8 +7,13 @@
...
 
 class RunParameters(MutableMapping):
-  def __init__(self, run_context, fixed_items):
-    _set(self, "__items", dict(fixed_items))
+  def __init__(self, run_context, fixed_items, default_items=None):
+    if default_items:
+      items = dict(default_items)
+      items.update(dict(fixed_items))
+    else:
+      items = dict(fixed_items)
```
[the remark this answers is not in the document]
Comment (yuhong-sigopt) [reply sigopt/run_params.py:15]: I follow the old code `_set(self, "__items", dict(fixed_items))`, I will change it.

**Review by taylorjacklespriggs — APPROVED** [review]
